File: js/attack.js
const fs = require("fs");
const child_process = require('child_process');
const os = require('os').type();
var arguments = process.argv[2];
file_name = arguments;
const source = fs.readFileSync(file_name, {encoding: 'utf8'});
lines = [];
if (os === "'Windows_NT'") {
    lines = source.split("\r\n");
} else {
    lines = source.split("\n");
}
const regex = lines[0];
const modifiers = lines[1];
const model = lines[2];
let attackBeans = [];
let i = 3;
while (i < lines.length - 1) {
    let attacks = lines[i + 1].split("IOS_AC_CN");
    let o = createAttackBean(lines[i], attacks[0], attacks[1], attacks[2]);
    attackBeans.push(o);
    i += 2
}

for (j = 0, len = attackBeans.length; j < len; j++) {
    if (attackBeans[j].type.toString().trim() === "EXPONENT") {
        let times = 1000;
        let duration = get_duration(attackBeans[j], times, 150);
        if (duration >= 150) {
            attackBeans[j].is_attack_success = true;
            attackBeans[j].times = times;
            attackBeans[j].duration = duration;
            break;
        }
    } else if (attackBeans[j].type.toString().trim() === "POLYNOMIAL") {
        let times = 100000;
        let duration = get_duration(attackBeans[j], times, 1000);
        if (duration >= 1000) {
            attackBeans[j].is_attack_success = true;
            attackBeans[j].times = times;
            attackBeans[j].duration = duration;
        }
    } else if (attackBeans[j].type.toString().trim() === "EXPONENT_OR_POLYNOMIAL") {
        let times = 1000;
        let duration = get_duration(attackBeans[j], times, 150);
        if (duration >= 150) {
            attackBeans[j].is_attack_success = true;
            attackBeans[j].times = times;
            attackBeans[j].duration = duration;
        }
        if (!attackBeans[j].is_attack_success) {
            times = 100000;
            let duration = get_duration(attackBeans[j], times, 1000);
            if (duration >= 1000) {
                attackBeans[j].is_attack_success = true;
                attackBeans[j].times = times;
                attackBeans[j].duration = duration;
            }
        }
    }
    if (model.toString().trim() === "s" && attackBeans[j].is_attack_success) {
        break
    }
}

let results = "";
for (j = 0, len = attackBeans.length; j < len; j++) {
    let bean = attackBeans[j];
    let line = bean.is_attack_success + "IOS_AC_CN" + bean.times + "IOS_AC_CN" + bean.duration;
    results = results + line + "\n";
}
fs.writeFileSync(file_name.replace(".txt", "_result.txt"), results);


function createAttackBean(type, prefix, repeat, suffix) {
    let o = {};
    o.type = type;
    o.prefix = prefix;
    o.repeat = repeat;
    o.suffix = suffix;
    o.is_attack_success = false;
    o.duration = 0;
    o.times = 0;
    return o;
}

function get_duration(bean, times, timeout) {
    let text = regex + "\n" + modifiers + "\n" + bean.prefix + "\n" + bean.repeat + "\n" + bean.suffix + "\n" + times + "\n";
    let temp = "js/temp_" + Date.now() + ".txt";
    fs.writeFileSync(temp, text);
    let duration = 0;
    var time = Date.now();
    try {
        duration = Number(child_process.execSync('node js/support.js ' + temp, {timeout: timeout}).toString().trim());
    } catch (e) {
        duration = Date.now() - time;
    }
    fs.unlinkSync(temp);
    return duration;
}

File: python/attack.py
import multiprocessing
import time
import re
import signal
import sys



class attack_bean:
    def __init__(self, type, prefix, repeat, suffix):
        self.type = type
        self.prefix = prefix
        self.repeat = repeat
        self.suffix = suffix
        self.is_attack_success = False
        self.duration = 0
        self.times = 0


def time_out(b, c):
    raise TimeoutError


def search_with_timeout(pipe, pattern, attack_str, timeout):
    signal.signal(signal.SIGALRM, time_out)
    signal.setitimer(signal.ITIMER_REAL, timeout)
    start = time.time()
    try:
        re.search(pattern, attack_str, flags=0)
        duration = int((time.time() - start) * 1000)
        pipe.send(duration)
    except TimeoutError:
        duration = int((time.time() - start) * 1000)
        pipe.send(duration)


def get_duration(bean, times, timeout):
    string = (bean.prefix + bean.repeat * times + bean.suffix).replace("\\n", "\n").replace("\\f", "\f").replace(
        "\\v",
        "\v").replace(
        "\\r", "\r").replace("\\t", "\t").replace("\\\\", "\\")

    sub_pipe = multiprocessing.Pipe()
    p = multiprocessing.Process(target=search_with_timeout, args=(sub_pipe[0], regex, string, timeout))
    p.start()
    p.join()
    return sub_pipe[1].recv()

if __name__ == '__main__':
    file_name = sys.argv[1]

    attack_file = open(file_name)
    lines = attack_file.readlines()
    regex = lines[0].strip("\n")
    model = lines[1].strip("\n")
    attack_beans = []
    length = lines.__len__()
    i = 2
    while i < length:
        type = lines[i].rstrip("\n")
        attacks = lines[i + 1].rstrip("\n").split("IOS_AC_CN")
        attack_beans.append(attack_bean(type, attacks[0], attacks[1], attacks[2]))
        i += 2

    for bean in attack_beans:
        type = bean.type
        if type == "EXPONENT":
            times = 10
            while times < 515:
                duration = get_duration(bean, times, 0.15)
                if duration >= 150:
                    bean.is_attack_success = True
                    bean.duration = duration
                    bean.times = times
                    break
                times += 50
        if type == "POLYNOMIAL":
            times = 10000
            while times < 320000:
                duration = get_duration(bean, times, 1)
                if duration >= 1000:
                    bean.is_attack_success = True
                    bean.duration = duration
                    bean.times = times
                    break
                times *= 2
        if type == "EXPONENT_OR_POLYNOMIAL":
            times = 10
            while times < 515:
                duration = get_duration(bean, times, 0.15)
                if duration >= 150:
                    bean.is_attack_success = True
                    bean.duration = duration
                    bean.times = times
                    break
                times += 50
            if not bean.is_attack_success:
                times = 10000
                while times < 320000:
                    duration = get_duration(bean, times, 1)
                    if duration >= 1000:
                        bean.is_attack_success = True
                        bean.duration = duration
                        bean.times = times
                        break
                    times *= 2
        if model == "s" and bean.is_attack_success:
            break

    result_file = open(file_name.replace(".txt", "_result.txt"), "w")
    for bean in attack_beans:
        line = str(bean.is_attack_success) + "IOS_AC_CN" + str(bean.times) + "IOS_AC_CN" + str(bean.duration) + "\n"
        result_file.write(line)
    result_file.close()
    print("python exit")